File: api/routers/sensitivity.py
"""
api/routers/sensitivity.py — R2: Server-side Parameter Sensitivity Analysis

POST /api/sensitivity
  Runs central-difference elasticity for all 14 backend-mappable params × 10 KPIs
  using the full Python DAG physics model (330ms for 28 evaluations).

Why server-side vs TypeScript:
  - No 28 separate HTTP round-trips (one request, 28 internal evaluations)
  - Full DAG model accuracy (vs simplified TS computeAll)
  - forkLength excluded — not a direct DAGRequest input field
"""
import time
import math
from fastapi import APIRouter
from pydantic import BaseModel
from typing import List, Optional

from .dag_analysis import DAGRequest
from ._physics import PARAM_META, KPI_META, eval_kpis, patch_request

router = APIRouter()


# ── Request ───────────────────────────────────────────────────────────────────

class SensitivityRequest(BaseModel):
    bike:          DAGRequest
    perturb_pct:   float = 2.0
    active_groups: Optional[List[str]] = None   # None = all three


# ── Response ──────────────────────────────────────────────────────────────────

class SensCell(BaseModel):
    elasticity: float
    raw_deriv:  float

class SensParamOut(BaseModel):
    id:    str
    label: str
    group: str
    unit:  str

class SensKPIOut(BaseModel):
    id:    str
    label: str
    unit:  str

class SensitivityResponse(BaseModel):
    params:        List[SensParamOut]
    kpis:          List[SensKPIOut]
    cells:         List[List[SensCell]]   # [paramIdx][kpiIdx]
    baseline_kpi:  List[float]
    baseline_param:List[float]
    perturb_pct:   float
    compute_ms:    float


# ── Endpoint ──────────────────────────────────────────────────────────────────

@router.post("/sensitivity", response_model=SensitivityResponse)
def sensitivity(req: SensitivityRequest) -> SensitivityResponse:
    t0 = time.perf_counter()

    allowed_groups = set(req.active_groups) if req.active_groups else {'Geometry', 'Suspension', 'Chain'}
    params = [p for p in PARAM_META if p['group'] in allowed_groups]
    kpis   = KPI_META

    # Baseline
    base_kpis   = eval_kpis(req.bike)
    baseline_kpi   = [base_kpis[k['id']] for k in kpis]
    baseline_param = []
    for p in params:
        field = p['field']
        val   = getattr(req.bike, field, None)
        baseline_param.append(float(val) if val is not None else 0.0)

    # Central difference per param
    cells: list[list[SensCell]] = []
    for pi, param in enumerate(params):
        pv = baseline_param[pi]
        delta = max(abs(pv) * (req.perturb_pct / 100.0), 1e-4)

        hi_kpis = eval_kpis(patch_request(req.bike, param['id'], pv + delta))
        lo_kpis = eval_kpis(patch_request(req.bike, param['id'], pv - delta))

        row: list[SensCell] = []
        for ki, kpi in enumerate(kpis):
            kpi_base = baseline_kpi[ki]
            kpi_hi   = hi_kpis[kpi['id']]
            kpi_lo   = lo_kpis[kpi['id']]

            raw_deriv  = (kpi_hi - kpi_lo) / (2 * delta)
            # E = (ΔK/K) / (ΔP/P) — dimensionless elasticity
            if kpi_base != 0 and pv != 0:
                elasticity = (raw_deriv * pv) / kpi_base
            else:
                elasticity = 0.0

            row.append(SensCell(
                elasticity = round(elasticity, 6) if math.isfinite(elasticity) else 0.0,
                raw_deriv  = round(raw_deriv, 6)  if math.isfinite(raw_deriv)  else 0.0,
            ))
        cells.append(row)

    compute_ms = (time.perf_counter() - t0) * 1000.0

    return SensitivityResponse(
        params         = [SensParamOut(**{k: p[k] for k in ('id','label','group','unit')}) for p in params],
        kpis           = [SensKPIOut(**{k: kp[k] for k in ('id','label','unit')}) for kp in kpis],
        cells          = cells,
        baseline_kpi   = baseline_kpi,
        baseline_param = baseline_param,
        perturb_pct    = req.perturb_pct,
        compute_ms     = round(compute_ms, 1),
    )

File: chassis-workbench/src/engine/sensitivity.ts
/**
 * sensitivity.ts — Parameter Sensitivity / Elasticity Engine (R2)
 *
 * For each input parameter, perturbs ±δ% and computes central-difference
 * elasticity:  E_ij = (∂KPI_j / KPI_j) / (∂param_i / param_i)
 *
 * Dimensionless — directly comparable across all param/KPI unit pairs.
 * E = +1.0 means "1% param increase → 1% KPI increase".
 * E = −2.0 means "1% param increase → 2% KPI decrease".
 */

import { ComputeAllInput, ComputeAllResult } from './types';
import { computeAll } from './computeAll';

// ── Parameter + KPI descriptors ──────────────────────────────────────────────

export type SensGroup = 'Geometry' | 'Suspension' | 'Chain';

export interface SensParam {
  id:         string;
  label:      string;
  group:      SensGroup;
  unit:       string;
  getValue:   (input: ComputeAllInput) => number;
  patchValue: (input: ComputeAllInput, v: number) => ComputeAllInput;
}

export interface SensKPI {
  id:       string;
  label:    string;
  unit:     string;
  getValue: (r: ComputeAllResult, input: ComputeAllInput) => number;
}

export interface SensCell {
  elasticity: number;   // (ΔKpi/Kpi) / (ΔParam/Param) — dimensionless
  rawDeriv:   number;   // ΔKpi / ΔParam — in KPI-units per param-unit
}

export interface SensitivityResult {
  params:       SensParam[];
  kpis:         SensKPI[];
  cells:        SensCell[][];   // [paramIdx][kpiIdx]
  baselineKPI:  number[];
  baselineParam:number[];
  perturbPct:   number;
  computeMs:    number;
}

// ─────────────────────────────────────────────────────────────────────────────

export const SENS_PARAMS: SensParam[] = [
  // ── Geometry ────────────────────────────────────────────────────────────────
  {
    id: 'headAngle', label: 'Head Angle', group: 'Geometry', unit: '°',
    getValue:   i => i.geometry.headAngle,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, headAngle: v } }),
  },
  {
    id: 'forkOffset', label: 'Fork Offset', group: 'Geometry', unit: 'mm',
    getValue:   i => i.geometry.forkOffset,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, forkOffset: v } }),
  },
  {
    id: 'forkLength', label: 'Fork Length', group: 'Geometry', unit: 'mm',
    getValue:   i => i.geometry.forkLength,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, forkLength: v } }),
  },
  {
    id: 'wheelbase', label: 'Wheelbase', group: 'Geometry', unit: 'mm',
    getValue:   i => i.geometry.wheelbase,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, wheelbase: v } }),
  },
  {
    id: 'swingarmLen', label: 'Swingarm Len', group: 'Geometry', unit: 'mm',
    getValue:   i => i.geometry.swingarmLength,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, swingarmLength: v } }),
  },
  {
    id: 'pivotHeight', label: 'Pivot Height', group: 'Geometry', unit: 'mm',
    getValue:   i => i.geometry.swingarmPivotHeight,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, swingarmPivotHeight: v } }),
  },
  {
    id: 'pivotX', label: 'Pivot X', group: 'Geometry', unit: 'mm',
    getValue:   i => i.geometry.swingarmPivotX,
    patchValue: (i, v) => ({ ...i, geometry: { ...i.geometry, swingarmPivotX: v } }),
  },
  // ── Suspension ──────────────────────────────────────────────────────────────
  {
    id: 'springRateF', label: 'Spring Rate F', group: 'Suspension', unit: 'N/mm',
    getValue:   i => i.suspension.springRateFront,
    patchValue: (i, v) => ({ ...i, suspension: { ...i.suspension, springRateFront: v } }),
  },
  {
    id: 'springRateR', label: 'Spring Rate R', group: 'Suspension', unit: 'N/mm',
    getValue:   i => i.suspension.springRateRear,
    patchValue: (i, v) => ({ ...i, suspension: { ...i.suspension, springRateRear: v } }),
  },
  {
    id: 'motionRatioF', label: 'Motion Ratio F', group: 'Suspension', unit: '',
    getValue:   i => i.suspension.motionRatioFront,
    patchValue: (i, v) => ({ ...i, suspension: { ...i.suspension, motionRatioFront: v } }),
  },
  {
    id: 'motionRatioR', label: 'Motion Ratio R', group: 'Suspension', unit: '',
    getValue:   i => i.suspension.motionRatioRear,
    patchValue: (i, v) => ({ ...i, suspension: { ...i.suspension, motionRatioRear: v } }),
  },
  {
    id: 'unsprungF', label: 'Unsprung F', group: 'Suspension', unit: 'kg',
    getValue:   i => i.suspension.unsprungFront,
    patchValue: (i, v) => ({ ...i, suspension: { ...i.suspension, unsprungFront: v } }),
  },
  {
    id: 'unsprungR', label: 'Unsprung R', group: 'Suspension', unit: 'kg',
    getValue:   i => i.suspension.unsprungRear,
    patchValue: (i, v) => ({ ...i, suspension: { ...i.suspension, unsprungRear: v } }),
  },
  // ── Chain / Transmission ────────────────────────────────────────────────────
  {
    id: 'frontSprocket', label: 'Front Sprocket', group: 'Chain', unit: 'T',
    getValue:   i => i.chain.frontSprocket,
    patchValue: (i, v) => ({ ...i, chain: { ...i.chain, frontSprocket: v } }),
  },
  {
    id: 'rearSprocket', label: 'Rear Sprocket', group: 'Chain', unit: 'T',
    getValue:   i => i.chain.rearSprocket,
    patchValue: (i, v) => ({ ...i, chain: { ...i.chain, rearSprocket: v } }),
  },
];

export const SENS_KPIS: SensKPI[] = [
  { id: 'trail',     label: 'Trail',    unit: 'mm',   getValue: r => r.geometry.trail },
  { id: 'frontPct',  label: 'F/R %',   unit: '%',    getValue: r => r.cog.frontPercent },
  { id: 'cogH',      label: 'CoG H',   unit: 'mm',   getValue: r => r.cog.Y_cg },
  { id: 'antiSquat', label: 'AS%',     unit: '%',    getValue: r => r.antiSquat.antiSquatPercent },
  { id: 'natFreqF',  label: 'Freq F',  unit: 'Hz',   getValue: r => r.suspension.natFreqFront },
  { id: 'natFreqR',  label: 'Freq R',  unit: 'Hz',   getValue: r => r.suspension.natFreqRear },
  { id: 'sagF',      label: 'Sag% F',  unit: '%',    getValue: r => r.suspension.sagPercentFront },
  { id: 'sagR',      label: 'Sag% R',  unit: '%',    getValue: r => r.suspension.sagPercentRear },
  { id: 'wrF',       label: 'WR F',    unit: 'N/mm', getValue: r => r.suspension.wheelRateFront },
  { id: 'wrR',       label: 'WR R',    unit: 'N/mm', getValue: r => r.suspension.wheelRateRear },
];

// ─────────────────────────────────────────────────────────────────────────────

/**
 * Run central-difference sensitivity analysis.
 * @param input        Current bike input
 * @param perturbPct   Perturbation size as % of each param's nominal value (default 2%)
 * @param activeGroups Groups to include; all three by default
 */
export function computeSensitivity(
  input: ComputeAllInput,
  perturbPct  = 2,
  activeGroups: Set<SensGroup> = new Set(['Geometry', 'Suspension', 'Chain']),
): SensitivityResult {
  const t0 = performance.now();

  const params = SENS_PARAMS.filter(p => activeGroups.has(p.group));

  // Baseline
  const baseResult    = computeAll(input);
  const baselineKPI   = SENS_KPIS.map(k => k.getValue(baseResult, input));
  const baselineParam = params.map(p => p.getValue(input));

  const cells: SensCell[][] = params.map((param, pi) => {
    const pv = baselineParam[pi];
    // δ = perturbPct% of nominal, with a floor so we never perturb by 0
    const δ  = Math.max(Math.abs(pv) * (perturbPct / 100), 1e-4);

    const hiResult = computeAll(param.patchValue(input, pv + δ));
    const loResult = computeAll(param.patchValue(input, pv - δ));

    return SENS_KPIS.map((kpi, ki) => {
      const kpiBase = baselineKPI[ki];
      const kpiHi   = kpi.getValue(hiResult, input);
      const kpiLo   = kpi.getValue(loResult, input);

      const rawDeriv  = (kpiHi - kpiLo) / (2 * δ);
      // E = (ΔK/K) / (ΔP/P)  — dimensionless elasticity
      const elasticity = (kpiBase !== 0 && pv !== 0)
        ? (rawDeriv * pv) / kpiBase
        : 0;

      return { elasticity: isFinite(elasticity) ? elasticity : 0, rawDeriv };
    });
  });

  return {
    params, kpis: SENS_KPIS, cells,
    baselineKPI, baselineParam,
    perturbPct,
    computeMs: performance.now() - t0,
  };
}
